File: Task0/Subject_solution.py
import random as rand


class Subject_solution:
    job_lookup = {}

    def __init__(self, permutation, makespan):
        self.permutation = permutation
        self.makespan = makespan

    def copy(self):
        return Subject_solution(self.permutation.copy(), self.makespan)

    @classmethod
    def set_lookup(cls, subject):
        cls.job_lookup = {job.id: job for job in subject}

    @classmethod
    def from_ids(cls, child_ids, machines):
        """Alternative constructor — build from a list of ids"""
        perm = [cls.job_lookup[id] for id in child_ids]
        ms = solution(perm, machines)[-1]
        return cls(perm, ms)

    @property
    def ids(self):
        return [job.id for job in self.permutation]

    def __repr__(self):
        return f"Subject_solution(makespan={self.makespan}, order={self.ids})"


# Subject_solution = S
# S[i] = (makespan=2103, order=[17, 8, ..., 11])
# S[i][1] = [17, 8, ..., 11]


def solution(jobs, machines):
    time = [0] * machines
    for jb in jobs:
        time[0] += jb.times[0]
        for i in range(1, machines):
            time[i] = max(time[i], time[i - 1]) + jb.times[i]
    return time


def swap(subject, machines):
    n = len(subject.permutation)
    i, j = rand.sample(range(n), 2)

    subject.permutation[i], subject.permutation[j] = (
        subject.permutation[j],
        subject.permutation[i],
    )

    subject.makespan = solution(subject.permutation, machines)[-1]


def ox(parent1, parent2, machines):
    child_ids = [None for _ in range(len(parent1))]
    i = rand.randint(0, len(parent1) - 1)
    j = rand.randint(0, len(parent1) - 1)
    if i > j:
        i, j = j, i
    child_ids[i:j] = parent1[i:j]
    for gene in parent2:
        if gene not in child_ids:
            for k in range(len(child_ids)):
                if child_ids[k] is None:
                    child_ids[k] = gene
                    break
    return Subject_solution.from_ids(child_ids, machines)

File: Task0/CppVersion/Subject_solution.cpp
#include "Subject_solution.h"

std::unordered_map<int, Job> Subject_solution::job_lookup;

void Subject_solution::set_lookup(const std::vector<Job>& subject) {
    for (const Job& job : subject) {
        job_lookup[job.id] = job;
    }
}

Subject_solution::Subject_solution(const std::vector<Job>& permutation)
    : permutation(permutation) {
    compute_makespan();
}
Subject_solution Subject_solution::from_ids(const std::vector<int>& child_ids) {
    std::vector<Job> perm;
    for (int id : child_ids) {
        perm.push_back(job_lookup[id]);
    }
    return Subject_solution(perm);  // constructor auto-computes makespan
}

Subject_solution Subject_solution::copy() const {
    return Subject_solution(permutation, makespan);
}

std::vector<int> Subject_solution::get_ids() const {
    std::vector<int> result;
    for (const Job& job : permutation) {
        result.push_back(job.id);
    }
    return result;
}

void Subject_solution::swap_jobs(std::mt19937& gen) {
    std::uniform_int_distribution<int> dist(0, (int)permutation.size() - 1);
    int i = dist(gen);
    int j = dist(gen);
    std::swap(permutation[i], permutation[j]);
    compute_makespan();
}

Job Subject_solution::pop_front() {
    Job first = permutation[0];
    permutation.erase(permutation.begin());
    return first;
}
void Subject_solution::remove_job(const Job& job) {
    permutation.erase(
        std::find_if(permutation.begin(), permutation.end(),
                     [&](const Job& j) { return j.id == job.id; }));
}
void Subject_solution::append_job(const Job& job) {
    permutation.push_back(job);
    compute_makespan();  // always stays up to date
}
std::string Subject_solution::repr() const {
    std::string result =
        "Subject_solution(makespan=" + std::to_string(makespan) + ", order=[";
    std::vector<int> id_list = get_ids();
    for (int i = 0; i < (int)id_list.size(); i++) {
        result += std::to_string(id_list[i]);
        if (i < (int)id_list.size() - 1) result += ", ";
    }
    result += "])";
    return result;
}